alf/pretrained_models/model_adapters.py

```diff
@@ -0,0 +1,318 @@
+            scaling /= self._r
+        return scaling
+
+    def _adapt(self, m: nn.Module):
+        """Adapt a module *in place*.
+
+        After this, ``m.forward()`` will be computed with the adapter weights.
+        """
+        assert not hasattr(m, '_forward0'), (
+            "The module has already been adapted! You need to first remove the "
+            "adapter.")
+        forward0 = m.forward
+        adapter_forward = self.forward
+
+        def _unmerged_forward(self, input):
+            """For unmerged forward, the base model and adapter forward separately,
+            and their results are weighted combined.
+
+            This forward way should be used for training.
+            """
+            return forward0(input) + adapter_forward(input)
+
+        m.forward = types.MethodType(_unmerged_forward, m)
+        m._unmerged_forward = m.forward
+        m._forward0 = forward0
+
+    def merge(self):
+        """Merge the adapter and base module weights.
+
+        This operation will change the base module weight in place and restore
+        its original ``forward()``. It should only be called in the inference mode.
+        """
+        if self._merged:
+            return
+        m = self._m[0]
+        m.weight.data.add_(self._adapter_weight())
+        m.forward = m._forward0
+        self._merged = True
+
+    def unmerge(self):
+        """Unmerge the adapter weights. The base module is still adapted, and
+        the adapter weights can be trained after unmerge.
+        """
+        if not self._merged:
+            return
+        m = self._m[0]
+        m.weight.data.add_(-self._adapter_weight())
+        m.forward = m._unmerged_forward
+        self._merged = False
+
+    def detach(self):
+        """Detach from and recover the base module.
+
+        Note that this operation is *irreversible*. Once detached, there is no
+        way to add the adapter back to the module.
+
+        We still keep the adapter weights after detachment.
+        """
+        if self._merged:
+            self.unmerge()
+        m = self._m[0]
+        m.forward = m._forward0
+        del m._forward0
+        del m._unmerged_forward
+
+    @classmethod
+    def can_adapt(cls, m: nn.Module) -> bool:
+        """Check if the adapter class can adapt a given module.
+        """
+        raise NotImplementedError()
+
+    @classmethod
+    def _decompose_weight_dims(cls, m: nn.Module) -> Tuple[int]:
+        """Return two (reshaped) dimensions ``(nrows, ncols)`` so that
+        ``nrows*ncols == np.prod(m.weight.shape)``.
+
+        The reinterpreted shape will be used to create LoRA weights ``self._wA``
+        and ``self._wB``.
+
+        Returns:
+            tuple: a pair of ints ``(nrows, ncols)``.
+        """
+        raise NotImplementedError()
+
+    def forward(self, input):
+        raise NotImplementedError()
+
+
+@alf.configurable
+class EmbeddingAdapter(LoRA):
```

Comment:
What about putting those concrete implementations in a file (module) `lora.py`, so that when used it looks like `lora.EmbeddingAdapter`? This improves readability slightly, in case there are other adapters.

In factor it looks like this `model_adapters.py` looks like it is dedicate for LoRA at this moment.

Reply:
> What about putting those concrete implementations in a file (module) `lora.py`, so that when used it looks like `lora.EmbeddingAdapter`? This improves readability slightly, in case there are other adapters.
> 
> In factor it looks like this `model_adapters.py` looks like it is dedicate for LoRA at this moment.

I've put adapters into model_adapters/lora.py